app/models/phone_number_authentication.rb

```diff
@@ -32,6 +36,30 @@ def mark_as_logged_in
     save
   end
 
+  def unlock
+    update!(locked_at: nil, failed_attempts: 0)
+  end
+
+  def in_lockout_period?
+    locked_at >= lockout_time.ago
+  end
+
+  def lockout_time
+    USER_AUTH_LOCKOUT_IN_MINUTES.minutes
+  end
+
+  def track_failed_attempt
+    increment!(:failed_attempts)
+    if failed_attempts >= USER_AUTH_MAX_FAILED_ATTEMPTS
+      update!(locked_at: Time.current)
+    end
+  end
+
+  def minutes_left_on_lockout
+    minutes_left = (lockout_time - (Time.current - locked_at)) / 1.minute
+    minutes_left = minutes_left.round
+  end
```

Comment:
Now that we've moved several pieces of this to the model, can we add the necessary specs to the corresponding model spec, so we know they work in isolation? The `phone_number_authentication/authenticate_spec` could either remain unchanged, or stub its interactions with the `authentication` object. I'll let ya take a judgment call on that.

Reply:
👍 added specs